### ISSUE TITLE
Enable CLA Bot to work in Gitlab akin to how it works in Github

### ISSUE THREAD

**Issue by brooklynrob** [finos/open-developer-platform#15]
Enable CLA Bot to work in Gitlab akin to how it works in Github. 

This is time sensitive for the PURE/Alloy pilot and ideally would be in place by the end of January 2020. Here's why ... PURE models will be stored initially on a hosted instance of GitLab Ultimate, later on Gitlab.com (https://gitlab.com/finosfoundation), therefore the app must be compatible with both configurations. Edits to models will be git commits and as such will require CLAs for modelers. 

More generally having this in place will give us flexibility within FINOS to support both Github and Gitlab

Sub-items:
- [x] Confirm with GS if Kiitos uses commits or Merge Requests to update code. If commits are used, what is the workflow to notify issues, since no MR checks . /CC 
- [x] Implement a group-level configuration in GitLab, see https://github.com/ScottLogic/gitlab-cla-bot/pull/49 ; I deployed and tested it already, seems to work as expected (test is hosted on https://gitlab.alloy.finos.org/test/clabot-config )
- [x] Generate a list of whitelisted GitHub IDs from our internal metadata
- [ ] Push the GitHub ID whitelist into https://gitlab.alloy.finos.org/Alloy-Pilot/clabot-config
- [ ] Implement a `useGitHubIDs` toggle in the CLA bot that informs the bot that contributors whitelist is now populated with GitHub IDs, instead of GitLab usernames

Technical discussion on https://github.com/finos/cla-bot/issues/182

**Comment by maoo** [comment]
Email sent out to GitLab to ask for help, they probably have lots of experience with this type of projects and can join efforts with FINOS and ScottLogic to make this happen. /CC @ColinEberhardt

**Comment by maoo** [comment]
@brooklynrob - FYI - took the freedom to update issue description and add the requirement for the app to also work on GitLab Ultimate.

**Comment by ColinEberhardt** [comment]
HI @maoo - just FYI, we have a small team who can look into this. They'll get started shortly :-)

**Comment by maoo** [comment]
Great news, TY @ColinEberhardt . I would love to be involved in the initiative, as it will give me the opportunity to deploy and test it against gitlab.com/finosfoundation and our internal GitLab hosted instance, making it easier to anticipate blocking issues.

One of the things I haven't figured out yet is whether we need to build a new contributors.json with GitLab IDs, or if it's possible/sufficient to know the GitHub ID of a GitLab user, assuming he/she logged in using GitHub Oauth SSO (or both options).

Probably first step is to define technical requirements (in the issue description above), as a result of an initial feasibility study; I'll leave that up to you and team, but I'm more than happy to help!

**Comment by ColinEberhardt** [comment]
Cool, we'll keep you in the loop. I think before we look at technical requirements we do need to check technical feasibility - does GitLab have sufficient APIs and webhooks, and an app deployment mechanism to support something similar to cla-bot? The answer is probably 'yes', but we do need to check!

**Comment by mcleo-d** [comment]
@ColinEberhardt - Let me know if additional team members need to be added to the Scott Logic CCLA and I'll help onboard them for you ... 👍

Also ... we need to make a decision on whether the GitLab version of CLA Bot lives in the FINOS org on GitHub or whether it needs to be closer to GitLab as part of the integration and testing requirements.

In my opinion, keeping everything together in GitHub is preferred, so we're not spreading ourselves too thin across systems and we don't need to duplicate CLA metadata. But, this might not be great for the dev team.

Feel free to add thoughts as appropriate 🤔

**Comment by jbjonesjr** [comment]
what's the api/implementation difference?

**Comment by ColinEberhardt** [comment]
> what's the api/implementation difference?

we don't know yet - should be able to look into this next week.

**Comment by mcleo-d** [comment]
@ColinEberhardt - Are you happy to add an update to the ODP 29/01 agenda to talk through GitLab similarities to GitHub and the approach taken to port the CLABot across for PureAlloy?

It'll be great to learn from team experiences. 

https://finosfoundation.atlassian.net/wiki/spaces/FDX/pages/1300332551/2020-01-29+ODP+WG+Meeting+notes

**Comment by ColinEberhardt** [comment]
Yes, happy to update on the 29th. Although, by that point, I'm sure we'll have some working code in a public repo.

**Comment by mcleo-d** [comment]
The following update was supplied by @ColinEberhardt on **Thursday 30th January**.

### We have:
- created a bot that integrates via manual webhook integration
- created CI deployment for AWS (still needs a bit of work for S3 bucket configuration)
- the bot uses a GitLab identity and public access token - i.e. the bot is a user account
- cla-bot config is taken from the master branch of the project root
- MRs are checked via webhook, with commits interrogated.
- CLA violations are reported

Basically, it is fully functional. There are some features of the current bot that are not yet implemented (summon via comment, org-based config), however, it feels like a solid base.

I've asked the team to tidy up, fully test, deploy and open source their work so far. Following that we can look at additional feature prioritisation.

**Comment by brooklynrob** [comment]
Today was the target date to open up Alloy to modelers (i.e., the day by which this functionality would need to be in place in alloy.finos.org/studio). Due to several delays on both sides, we need to push that date, likely to next week (cc @beekemarie) but nonetheless it would be good to know how this (CLA Bot into Gitlab) is going, what's left to do, and - critically - when this will be done. @ColinEberhardt @mcleo-d @gs-bracej  @maoo

**Comment by brooklynrob** [comment]
@maoo @ColinEberhardt Any update on the CLA Bot and it's set up in the pilot instance of Alloy (github.alloy.finos.org)? Is it good to go? (CC @gs-bracej @beekemarie)

**Comment by mcleo-d** [comment]
Hi @ColinEberhardt - Is it possible to loop the CLA Bot engineers into this thread so we can start planning bot integration with GitLab? 

I'll also reach out to you over email. 

James.

**Comment by ColinEberhardt** [comment]
We've just made the cla-bot repo public: https://github.com/ScottLogic/gitlab-cla-bot

I think it gives you all the functionality you need to run on GitLab now. You can get a decent overview of the functionality from looking at the test specs:

https://github.com/ScottLogic/gitlab-cla-bot/blob/master/spec/indexSpec.js

There is also a project board here:

https://github.com/ScottLogic/gitlab-cla-bot/projects/1

They are currently working on a few tidy-up tasks for an MVP release. There are a few remaining features detailed here that have not been implemented:

https://github.com/ScottLogic/gitlab-cla-bot/issues?q=is%3Aissue+is%3Aopen+label%3AFinos

But hopefully it is sufficiently feature-rich in its current form.

**Comment by mcleo-d** [comment]
@ColinEberhardt - Do you have a Scott Logic developer who can help with the integration, configuration and test of the CLA Bot project?

Ideally, it'll be good to give them access to the Alloy GitLab with @maoo so they can do a full end-to-end integration and test.

Thanks so much 👍 

James.

**Comment by mcleo-d** [comment]
I've just raised the following issue on the Scott Logic GitLab CLA bot backlog to ask for help integrating the project into the Pure/Alloy version of GitLab ... https://github.com/ScottLogic/gitlab-cla-bot/issues/35

**Comment by maoo** [comment]
Thanks @mcleo-d for opening the issue!

I'm going to try deploying this into gitlab.alloy.finos.org ; @ColinEberhardt  - would be great to have a screensharing session with someone from the dev team, where we go through some testing together, wdyt?

**Comment by maoo** [comment]
Here's my progress so far - https://github.com/ScottLogic/gitlab-cla-bot/issues/35

I'm stuck on a 401 issue, any help would be greatly appreciated. TY!

**Comment by maoo** [comment]
Just updated description of this epic, pointing to the singular issues on the SL repo where source code is being developed.

**Comment by brooklynrob** [comment]
It is critical get an updated ETA on when this (CLA Bot integrated with the FINOS Alloy instance) will be tested, delivered, and ready. (CC @beekemarie @gs-bracej @pierredelen) as we have a need for this functionality to be in place to support the pilot. @maoo you are following up with @ColinEberhardt, correct, in addition to discussion here?

**Comment by ColinEberhardt** [comment]
I'm keen to talk too - I'm not sure exactly where we are at? I know our bot has pretty much all the functionality required now. @maoo let's chat!

**Comment by maoo** [comment]
Thanks @ColinEberhardt and @sehaswell for your time on Friday! Your suggestions were helpful to move forward, as I managed to:
- Implement a group-level configuration in GitLab, see https://github.com/ScottLogic/gitlab-cla-bot/pull/49 ; I deployed and tested it already, seems to work as expected (test is hosted on https://gitlab.alloy.finos.org/test/clabot-config )
- Managed to generate a list of whitelisted GitHub IDs from our internal metadata

In order to complete the setup of the GitLab CLA bot for FINOS, we still need to:
- Push the GitHub ID whitelist into https://gitlab.alloy.finos.org/Alloy-Pilot/clabot-config
- Implement a `useGitHubIDs` toggle in the CLA bot that informs the bot that contributors whitelist is now populated with GitHub IDs, instead of GitLab usernames

Issue description updated. I'll keep on working on it, deploying my forked version (http://github.com/maoo/gitlab-cla-bot) into our GitLab instance and submitting PRs to this repo for validation/feedback.

Thanks!

**Comment by maoo** [comment]
I've fixed all the outstanding issues (on https://github.com/maoo/gitlab-cla-bot) and enabled the CLA Bot on the `Alloy-Pilot` group (of projects) in gitlab.alloy.finos.org, details follow (access to gitlab.alloy.finos.org is restricted only to FINOS staff and pilot participants):

1. Pushing to GitLab clabot-config works, see https://gitlab.alloy.finos.org/test/clabot-config/commit/fce22beda23cc88baacb9bc941e4c939a982ab9c
2. GitHub ID verifier done, deployed and successfully tested on GitLab, see https://github.com/ScottLogic/gitlab-cla-bot/pull/50
3. CLA Bot integration is now configured at group level (on `Alloy-Pilot` and `Test` groups), so all Merge Requests across all projects within these groups will be scanned by the bot 

Tests are on https://gitlab.alloy.finos.org/test/mao-rob-test/merge_requests , CLA bot configuration is on https://gitlab.alloy.finos.org/test/clabot-config/blob/master/.clabot

Next steps:
- [ ] Final testing and review with @brooklynrob and @gs-bracej 
- [ ] Add demo to next ODP meeting
- [ ] Follow up on PRs - https://github.com/ScottLogic/gitlab-cla-bot/pull/49 and https://github.com/ScottLogic/gitlab-cla-bot/pull/50
- [ ] Restrict access to `clabot-config` project to only maintainers - @johncoghlan not sure if this is possible in GitLab, I couldn't find a configuration at project level to overwrite the view access to a project

**Comment by brooklynrob** [comment]
@maoo (@mcleo-d ) before we truly close this, could we set up time when you could walk me through where the message appears? Maybe you could walk through this at the ODP meeting? I think it would be useful for us to see this (i.e., see the screens) before we truly close this and consider it resolved.

**Comment by maoo** [comment]
@brooklynrob - sure, let's do this tomorrow during the ODP call, it's already in the agenda.

Meanwhile, screenshot are shared on https://github.com/finos/alloy/issues/82

**Comment by mcleo-d** [comment]
Hi @maoo

Just reviewing this story and have noticed the following aren't checked off in the acceptance criteria...

* Push the GitHub ID whitelist into https://gitlab.alloy.finos.org/Alloy-Pilot/clabot-config
* Implement a useGitHubIDs toggle in the CLA bot that informs the bot that contributors whitelist is now populated with GitHub IDs, instead of GitLab usernames

Can you confirm these items are either done or no longer completed?

Many thanks 🚀 

James.

**Comment by maoo** [comment]
Yes, they are both delivered. Thanks for checking!
